File: codingninjas_bfs_in_graph/bfs_queue_approach.cpp
// Time complexity: O(V + E), Space complexity: O(V + E) to create the graph
// O(V) for queue and visited boolean vector.
vector<int> bfsTraversal(int n, vector<vector<int>> &adj){
    vector<int> bfsOut;
    vector<int> graph[n];

    for (int i = 0; i < n; i++) {
        for (auto e : adj[i]) {
            if (e != 0) {
                graph[i].push_back(e);
            }
        }
    }
    vector<bool> visited(n);
    
    queue<int> q;
    q.push(0);

    while(!q.empty()) {
        int top = q.front();
        q.pop();

        if (!visited[top]) {
            bfsOut.push_back(top);
            visited[top] = true;
        }
        for (auto neigh : graph[top]) {
            if (!visited[neigh]) {
                q.push(neigh);
            }
        }
    }

    return bfsOut;
}

File: codingninjas_bfs_in_graph/bfs_queue_approach.py
from typing import List

def bfsTraversal(n: int, adj: List[List[int]]) -> List[int]:
    # write your code here
    bfs_trav = []
    visited = [False]*n
    graph = [[] for _ in range(n)]

    for node in range(n):
        for edge in adj[node]:
            if edge != 0:
                graph[node].append(edge)

    q = []
    q.append(0)
    visited[0] = True

    while len(q) != 0:
        top = q[0]
        q.pop(0)
        bfs_trav.append(top)

        for neigh in graph[top]:
            if not visited[neigh]:
                q.append(neigh)
                visited[neigh] = True
    
    return bfs_trav